### PR TITLE
[SPARK-40196][PYTHON][PS] Consolidate `lit` function with NumPy scalar in sql and pandas module

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Consolidate `lit` function with NumPy scalar in sql and pandas module.

Major changes:
- More accurate types for `lit` in sql with NumPy scalar input
- Replace `lit` in pandas with `lit` in sql

### Why are the changes needed?
As part of [SPARK-39405](https://issues.apache.org/jira/browse/SPARK-39405) for NumPy support in SQL.

That provides more precise and consistent Numpy support across PySpark moduels.

### Does this PR introduce _any_ user-facing change?
Yes. More accurate types for `lit`.

**FROM**
```py
>>> sdf.select(F.lit(np.int8(1)), F.lit(np.int16(1)), F.lit(np.int32(1)), F.lit(np.int64(1))).dtypes
[('1', 'int'), ('1', 'int'), ('1', 'int'), ('1', 'int')]
>>> sdf.select(F.lit(np.float32(1)), F.lit(np.float64(1))).dtypes
[('1.0', 'double'), ('1.0', 'double')]
```

**TO**
```py
>>> sdf.select(F.lit(np.int8(1)), F.lit(np.int16(1)), F.lit(np.int32(1)), F.lit(np.int64(1))).dtypes
[('CAST(1 AS TINYINT)', 'tinyint'), ('CAST(1 AS SMALLINT)', 'smallint'), ('CAST(1 AS INT)', 'int'), ('CAST(1 AS BIGINT)', 'bigint')]
>>> sdf.select(F.lit(np.float32(1)), F.lit(np.float64(1))).dtypes
[('CAST(1.0 AS FLOAT)', 'float'), ('CAST(1.0 AS DOUBLE)', 'double')]
>>> 
```

### How was this patch tested?
Unit tests.

### CODE REVIEW

**Comment by xinrong-meng** [comment]
It may be easier to review by commits :)

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/pandas/tests/test_spark_functions.py [hunk 1]
```diff
@@ -20,25 +20,12 @@
...
-    def test_lit(self):
-        self.assertTrue(spark_column_equals(SF.lit(np.int64(1)), F.lit(1).astype(LongType())))
-        self.assertTrue(spark_column_equals(SF.lit(np.int32(1)), F.lit(1).astype(IntegerType())))
-        self.assertTrue(spark_column_equals(SF.lit(np.int8(1)), F.lit(1).astype(ByteType())))
-        self.assertTrue(spark_column_equals(SF.lit(np.byte(1)), F.lit(1).astype(ByteType())))
-        self.assertTrue(
-            spark_column_equals(SF.lit(np.float32(1)), F.lit(float(1)).astype(FloatType()))
-        )
-        self.assertTrue(spark_column_equals(SF.lit(1), F.lit(1)))
+    def test_repeat(self):
```
Comment (xinrong-meng) [inline python/pyspark/pandas/tests/test_spark_functions.py:27]: This change is irrelevant to the PR proposal but is made to keep the `test_spark_functions` file.

We shall complete tests for `pyspark/pandas/spark/functions` here later.

**Comment by xinrong-meng** [comment]
CC @ueshin @zhengruifeng @itholic

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
Btw, I think we can further support `lit(np.ndarray)` like [`lit([...])`](https://github.com/apache/spark/pull/37722) in the future

**Comment by HyukjinKwon** [comment]
Can we do this ^ @itholic ?

**Comment by itholic** [comment]
Sure, just created ticket: https://issues.apache.org/jira/browse/SPARK-40462

Let me take a look

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by zhengruifeng** [comment]
this PR doesn't cover the newly added `groupby.quantile`:
```
starting python compilation test...
python compilation succeeded.

starting black test...
black checks passed.

starting flake8 test...
flake8 checks passed.

starting mypy annotations test...
annotations failed mypy checks:
python/pyspark/pandas/window.py:112: error: Module has no attribute "lit"  [attr-defined]
Found 1 error in 1 file (checked 340 source files)
1
```

fixed in https://github.com/apache/spark/pull/37908

**Comment by Yikun** [comment]
```
  test_repeat (pyspark.pandas.tests.test_spark_functions.SparkFunctionsTests) ... FAIL (0.052s)

======================================================================
FAIL [0.052s]: test_repeat (pyspark.pandas.tests.test_spark_functions.SparkFunctionsTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/spark/spark/python/pyspark/pandas/tests/test_spark_functions.py", line 28, in test_repeat
    self.assertTrue(spark_column_equals(SF.repeat(F.lit(1), 2), F.repeat(F.lit(1), 2)))
AssertionError: False is not true

----------------------------------------------------------------------
Ran 1 test in 8.471s
```

Still failed due to `test_repeat`, Initial invistigation:
```
F.repeat(F.lit(1), 2).__dict__
Out[3]: {'_jc': JavaObject id=o50}
SF.repeat(F.lit(1), 2).__dict__
Out[4]: {'_jc': JavaObject id=o54}
```

According to https://github.com/apache/spark/pull/37888#discussion_r971408190 , Looks like we need to remove this test?

```python
class SparkFunctionsTests(PandasOnSparkTestCase):
    def test_repeat(self):
        # TODO: Placeholder
        pass
```

**Comment by zhengruifeng** [comment]
@Yikun Let's comment this test for now, to unblock other PRs

**Comment by xinrong-meng** [comment]
Thank you @HyukjinKwon @zhengruifeng @Yikun for taking care of the merging!

**Pull request closed without merging**